lib/api/src/ticketAPI.cpp

```diff
@@ -0,0 +1,427 @@
+#include "irods/ticketAPI.hpp"
+
+#include <charconv>
+#include <string>
+#include <sstream>
+
+#include "irods/irods_random.hpp"
+#include "irods/objInfo.h"
+#include "irods/rcConnect.h"
+#include "irods/rcMisc.h"
+#include "irods/rodsError.h"
+#include "irods/rodsKeyWdDef.h"
+#include "irods/ticketAdmin.h"
+
+namespace irods::administration::ticket
+{
+    void make_ticket_name(char* newTicket)
+    {
+        const int ticket_len = 15;
+        // random_bytes must be (unsigned char[]) to guarantee that following
+        // modulo result is positive (i.e. in [0, 61])
+        unsigned char random_bytes[ticket_len];
+        irods::getRandomBytes(random_bytes, ticket_len);
+
+        const char characterSet[] = {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P',
+                                     'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f',
+                                     'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v',
+                                     'w', 'x', 'y', 'z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9'};
+
+        for (int i = 0; i < ticket_len; ++i) {
+            const int ix = random_bytes[i] % sizeof(characterSet);
+            newTicket[i] = characterSet[ix];
+        }
+        newTicket[ticket_len] = '\0';
+        printf("ticket:%s\n", newTicket);
+    }
+    int ticket_manager(RxComm& conn,
+                       std::string_view command,
+                       std::string_view ticket_IDentifier,
+                       std::string_view commandModifier,
+                       std::string_view commandModifier2,
+                       std::string_view commandModifier3,
+                       std::string_view commandModifier4,
+                       bool run_as_admin)
+    {
+        if (const int status = clientLogin(&conn); status != 0) {
+            throw USER_LOGIN_EXCEPTION(); // int error code -- client login didn't work
+            return 3;
+        }
+
+        ticketAdminInp_t ticketAdminInp{};
+
+        ticketAdminInp.arg1 = strdup(command.data());
+        ticketAdminInp.arg2 = strdup(ticket_IDentifier.data());
+        ticketAdminInp.arg3 = strdup(commandModifier.data());
+        ticketAdminInp.arg4 = strdup(commandModifier2.data());
+        ticketAdminInp.arg5 = strdup(commandModifier3.data());
+        ticketAdminInp.arg6 = strdup(commandModifier4.data());
+
+        if (run_as_admin) {
+            addKeyVal(&ticketAdminInp.condInput, ADMIN_KW, "");
```

Comment:
This will result in a memory leak.

You can defend against that by using `irods::at_scope_exit`. Here is an example:
```cpp
irods::at_scope_exit free_memory{[&ticketAdminInp] {
    clearKeyVal(&ticketAdminInp.condInput);
}};
```